maven-artifact/src/main/java/org/apache/maven/artifact/DefaultArtifact.java

```diff
@@ -314,7 +314,9 @@ public boolean equals(Object o) {
             return false;
         } else if (!a.getArtifactId().equals(artifactId)) {
             return false;
-        } else if (!a.getVersion().equals(version)) {
+        } else if (a.getVersion() != null && !a.getVersion().equals(version)) {
+            return false;
+        } else if (a.getVersion() == null && version != null) {
```

Comment:
I would suggest rewriting the whole logic in a single statement:
```
    return Objects.equals(a.getGroupId(), this.groupId)
        && Objects.equals(a.getArtifactId(), this.artifactId)
        && Objects.equals(a.getVersion(), this.version)
        && Objects.equals(a.getType(), this.type)
        && Objects.equals(a.getClassifier(), this.classifier);
```

Reply:
Agreed. I have revised this part.